loky/backend/_thread_pool_limiters.py

```diff
@@ -0,0 +1,443 @@
+
+    Return a dict dynamic_threadpool_size containing pairs `('clib': boolean)`
+    which are True if `clib` have been found and can be used to scale the
+    maximal number of threads dynamically.
+    """
+    wrapper = _get_wrapper(reload_clib)
+    return wrapper.set_thread_limits(limits, subset)
+
+
+def get_thread_limits(reload_clib=False):
+    """Return maximal thread number for threadpools in supported C-lib
+
+    Return a dictionary containing the maximal number of threads that can be
+    used in supported libraries or None when the library is not available. The
+    key of the dictionary are {}.
+
+    If `reload_clib` is `True`, first loop through the loaded libraries to
+    ensure that this function is called on all available libraries.
+    """.format(list(_CLibsWrapper.SUPPORTED_CLIBS.keys()))
+    wrapper = _get_wrapper(reload_clib)
+    return wrapper.get_thread_limits()
+
+
+class thread_pool_limits:
+    """Change the default number of threads used in thread pools.
+
+    This class can be used either as a function (the construction of this
+    object limits the number of threads) or as a context manager, in a `with`
+    block.
+
+    Parameters
+    ----------
+    limits : int or dict, (default=None)
+        Maximum number of thread that can be used in thread pools
+
+        If int, sets the maximum number of thread to `limits` for each C-lib
+        selected by `subset`.
+
+        If dict(supported_libraries: max_threads), sets a custom maximum number
+        of thread for each C-lib.
+
+        If None, does not do anything.
+
+    subset : string or None, optional (default="all")
+        Subset of C-libs to limit. Used only if `limits` is an int
+
+        "all" : limit all supported C-libs.
+
+        "blas" : limit only BLAS supported C-libs.
+
+        "openmp" : limit only OpenMP supported C-libs. It can affect the number
+                   of threads used by the BLAS C-libs if they rely on OpenMP.
+
+    .. versionadded:: 0.21
+
+    """
+    def __init__(self, limits=None, subset=None):
+
+        self.old_limits = get_thread_limits()
+        _set_thread_limits(limits=limits, subset=subset)
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, type, value, traceback):
+        self.unregister()
+
+    def unregister(self):
+        _set_thread_limits(limits=self.old_limits)
+
+
+def get_openblas_version(reload_clib=True):
+    """Return the OpenBLAS version
+
+    Parameters
+    ----------
+    reload_clib : bool, (default=True)
+        If `reload_clib` is `True`, first loop through the loaded libraries to
+        ensure that this function is called on all available libraries.
+
+    Returns
+    -------
+    version : string or None
+        None means OpenBLAS is not loaded or version < 0.3.4, since OpenBLAS
+        did not expose it's verion before that.
+    """
+    wrapper = _get_wrapper(reload_clib)
+    return wrapper.get_openblas_version()
```

Comment:
I would be in favor of a function that would return all the detected modules, e.g.:

```python
>>> get_threaded_libraries(api="all")
[
    {
        "module": <the ctypes module itself>,
        "api": "blas",  # or "openmb"
        "implementation": "openblas",
        "version": ...,
        "num_threads": 4,
    },
    {
        ...
    },
]
```

This would make it much easier for people reporting issues with parallelism with numpy, joblib, loky, scikit-learn and so on to give both them and us useful debug information to understand the state of their runtime environment.

It would also make it possible to write better/simpler tests for this module.

Reply:
Ok, I did a first round of refactoring to get this kind of API. Let me know what you think of it.
I will correct the other comments tomorrow.